### PR TITLE
chore(integer): add tests for default signed rotations/shifts

### PR DESCRIPTION
**Pull request by tmontaigu**
* [x] Tests for the changes have been added (for bug fixes / features)

### CODE REVIEW

File: tfhe/src/integer/server_key/radix_parallel/tests_signed.rs [hunk 1]
```diff
@@ -1474,6 +1518,342 @@ fn integer_signed_default_absolute_value(param: impl Into<PBSParameters>) {
...
+
+    sks.set_deterministic_pbs_execution(true);
+
+    let mut rng = rand::thread_rng();
+
+    let modulus = (cks.parameters().message_modulus().0.pow(NB_CTXT as u32) / 2) as i64;
+
+    assert!(modulus > 0);
+    assert!((modulus as u64).is_power_of_two());
+    let nb_bits = modulus.ilog2() + 1; // We are using signed numbers
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_signed.rs:1536]: how does that change the nb_bits and increases it ?

File: tfhe/src/integer/server_key/radix_parallel/tests_signed.rs [hunk 2]
```diff
@@ -1474,6 +1518,342 @@ fn integer_signed_default_absolute_value(param: impl Into<PBSParameters>) {
...
+            let mut shift = cks.encrypt(nb_bits);
+            sks.unchecked_add_assign(&mut shift, &tmp);
+            clear_shift += nb_bits;
+            assert!(!shift.block_carries_are_empty());
+
+            let ct_res = sks.left_shift_parallelized(&ct, &shift);
+            let dec_res: i64 = cks.decrypt_signed(&ct_res);
+            // When nb_bits is not a power of two
+            // then the behaviour is not the same
+            let mut nb_bits = modulus.ilog2();
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_signed.rs:1585]: can you elaborate a bit ?

**Review by IceTDrinker — COMMENTED** [review]
Same comments apply for the various tests

File: tfhe/src/integer/server_key/radix_parallel/tests_signed.rs (hunk 1 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/tests_signed.rs:1536]: Above, we computed the modulus as `let modulus = (cks.parameters().message_modulus().0.pow(NB_CTXT as u32) / 2) as i64;` so the unsigned modulus divided by two. So two have the number of bits we have to add that bit

File: tfhe/src/integer/server_key/radix_parallel/tests_signed.rs (hunk 2 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/tests_signed.rs:1585]: It comes from how barrel shifter is/are made (so thats also the case for unsigned radix)

If you want to shift an integer type with N bits, the barrel shift only consider the log2(N) bits from the value that gives the shift amount.

This works well when N is a power of two as it is the case for 8, 16, 32, 64, 128.
e.g for a u8 -> 8 bits -> log2(8) = 4, barrel shifter only consider bits in range 0..4 (so 3 bits), with 3 bits we can only give input shift in range 0..=7 -> great u8 has 8 bits
u16 -> 16 bits -> bits in range 0..4 considered -> possible shift values 0..15.
Basically the barrel shift naturally does `shift_amount % nb_bits` if N is a power of two.

for u12 (4 blocks of 3_3)  -> log2(12) = 3.58 -> we have to ceil() -> 4 -> bits in range 0..4 -> shift values 0..15 which is > 12.  Which means that the behaviour is not the same if the input shift value is not already in the good range

File: tfhe/src/integer/server_key/radix_parallel/tests_signed.rs (hunk 1 above)
Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix_parallel/tests_signed.rs:1536]: indeed

**Review by IceTDrinker — APPROVED** [review]

**Comment by tmontaigu** [comment]
Rebased on main to have fix for division

**Review by IceTDrinker — APPROVED** [review]

**Comment by IceTDrinker** [comment]
needs make fmt

**Review by IceTDrinker — APPROVED** [review]
